lib/engine/game/base.rb

```diff
@@ -192,11 +193,44 @@ def active_players
         @round.active_entities.map(&:owner)
       end
 
+      def process_undos(actions)
+        @undo_list = []
+        @last_known_undo = 0
+
+        stack = []
+        actions.each.with_index do |action, index|
+          # action_id's start at 1, so always add one to the index
+          if action['type'] == 'undo'
+            @last_known_undo = index + 1
+            @undo_list << stack.pop
+          elsif action['type'] == 'redo'
+            @last_known_undo = index + 1
+            @undo_list.pop
+          elsif action['type'] != 'message'
+            # Adding more types of action here will break existing games
+
+            stack.append(index + 1)
+          end
+        end
+      end
+
       def process_action(action)
```

Comment:
sorry to be picky, but i feel like process_action could just handle all of this without keep track of indicies

so first pass, you remove all the undos and redos that way process_action knows nothing about undos and redos except when they get them, they must do a clone

Reply:
I think the problem with this is when you come to the clone you'll need to restore the undos before being reprocessed.

Reply:
wouldn't you only pass the canonical data into the clone?

Reply:
yes, there's a secondary issue i need to find a solution for, since everything is passing action ids, there's a chance that if someone undos and does an action quickly enough two browsers can diverge in what the action id is.

Reply:
would the advisory locks i added help with this?